src/main/java/me/jellysquid/mods/sodium/client/gui/options/OptionPage.java

```diff
@@ -1,24 +1,65 @@
 package me.jellysquid.mods.sodium.client.gui.options;
 
 import com.google.common.collect.ImmutableList;
+import me.jellysquid.mods.sodium.client.SodiumClientMod;
 import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import org.embeddedt.embeddium.client.gui.EmbeddiumOptionsAPI;
+
+import java.util.ArrayList;
+import java.util.List;
 
 public class OptionPage {
+    public static final ResourceLocation DEFAULT_ID = new ResourceLocation(SodiumClientMod.MODID, "empty");
+
+    public static final ResourceLocation GENERAL = new ResourceLocation(SodiumClientMod.MODID, "general");
+    public static final ResourceLocation QUALITY = new ResourceLocation(SodiumClientMod.MODID, "quality");
+    public static final ResourceLocation PERFORMANCE = new ResourceLocation(SodiumClientMod.MODID, "performance");
+    public static final ResourceLocation ADVANCED = new ResourceLocation(SodiumClientMod.MODID, "advanced");
+
+    private final ResourceLocation id;
     private final Component name;
     private final ImmutableList<OptionGroup> groups;
     private final ImmutableList<Option<?>> options;
 
+    @Deprecated
     public OptionPage(Component name, ImmutableList<OptionGroup> groups) {
+        this(OptionPage.DEFAULT_ID, name, groups);
+        SodiumClientMod.logger().warn("Id must be specified in OptionPage '{}'", name.getString());
+    }
+
+    public OptionPage(ResourceLocation id, Component name, ImmutableList<OptionGroup> groups) {
+        this.id = id;
         this.name = name;
-        this.groups = groups;
 
-        ImmutableList.Builder<Option<?>> builder = ImmutableList.builder();
+        List<Option<?>> optionsBuilder = new ArrayList<>();
+        List<OptionGroup> groupsBuilder = new ArrayList<>();
 
-        for (OptionGroup group : groups) {
-            builder.addAll(group.getOptions());
+        for (int i = 0; i < groups.size(); i++) {
+            final OptionGroup group = groups.get(i);
+
+            if (group.getOptions().isEmpty()) {
+                continue; // ignore empty groups
+            }
+
+            if (!EmbeddiumOptionsAPI.consume(groupsBuilder, EmbeddiumOptionsAPI.customOptionGroups, group)) {
+                groupsBuilder.add(group);
+                optionsBuilder.addAll(group.getOptions());
+            } else {
+                if (groupsBuilder.get(i) == group) { // check if was added and add options too.
+                    optionsBuilder.addAll(group.getOptions());
+                }
+            }
         }
 
-        this.options = builder.build();
+        EmbeddiumOptionsAPI.consume(groupsBuilder, EmbeddiumOptionsAPI.customOptionGroups);
```

Comment:
Please revert the calls to `EmbeddiumOptionsAPI.consume` and just use the original Sodium logic in those areas for now; it will be done differently when I finish writing the new event system (in progress as we speak).